app/com/linkedin/drelephant/tuning/engine/MRExecutionEngine.java

```diff
@@ -0,0 +1,458 @@
+            CommonConstantsHeuristic.ParameterKeys.MAPPER_MEMORY_HADOOP_CONF.getValue())) {
+          memoryConstraints.put("CONTAINER_MEMORY", parameterConstraint);
+        }
+        if (parameterConstraint.tuningParameter.paramName.equals(
+            CommonConstantsHeuristic.ParameterKeys.MAPPER_HEAP_HADOOP_CONF.getValue())) {
+          memoryConstraints.put("CONTAINER_HEAP", parameterConstraint);
+        }
+      }
+      if (functionType.equals("reduce")) {
+        if (parameterConstraint.tuningParameter.paramName.equals(
+            CommonConstantsHeuristic.ParameterKeys.REDUCER_MEMORY_HADOOP_CONF.getValue())) {
+          memoryConstraints.put("CONTAINER_MEMORY", parameterConstraint);
+        }
+        if (parameterConstraint.tuningParameter.paramName.equals(
+            CommonConstantsHeuristic.ParameterKeys.REDUCER_HEAP_HADOOP_CONF.getValue())) {
+          memoryConstraints.put("CONTAINER_HEAP", parameterConstraint);
+        }
+      }
+    }
+    return memoryConstraints;
+  }
+
+  private void memoryParameterIPSO(String trigger, Map<String, TuningParameterConstraint> constraints,
+      List<Double> usageStats) {
+    logger.info(" IPSO for " + trigger);
+    Double usagePhysicalMemory = usageStats.get(UsageCounterSchema.USED_PHYSICAL_MEMORY.ordinal());
+    Double usageVirtualMemory = usageStats.get(UsageCounterSchema.USED_VIRTUAL_MEMORY.ordinal());
+    Double usageHeapMemory = usageStats.get(UsageCounterSchema.USED_HEAP_MEMORY.ordinal());
+    Double memoryMB =
+        applyContainerSizeFormula(constraints.get("CONTAINER_MEMORY"), usagePhysicalMemory, usageVirtualMemory);
+    applyHeapSizeFormula(constraints.get("CONTAINER_HEAP"), usageHeapMemory, memoryMB);
+  }
+
+  @Override
+  public void parameterOptimizerIPSO(List<AppResult> results, JobExecution jobExecution) {
+    Map<String, Map<String, Double>> previousUsedMetrics = extractParameterInformationIPSO(results);
+    List<TuningParameterConstraint> parameterConstraints = TuningParameterConstraint.find.where().
+        eq("job_definition_id", jobExecution.job.id).findList();
+    for (String function : functionTypes) {
+      logger.info(" Optimizing Parameter Space  " + function);
+      if (previousUsedMetrics.get(function)
+          .get(CommonConstantsHeuristic.UtilizedParameterKeys.MAX_PHYSICAL_MEMORY.getValue()) > 0.0) {
+        List<Double> usageStats = extractUsageParameter(function, previousUsedMetrics);
+        Map<String, TuningParameterConstraint> memoryConstraints =
+            filterMemoryConstraint(parameterConstraints, function);
+        memoryParameterIPSO(function, memoryConstraints, usageStats);
+      }
+    }
+  }
+
+  @Override
+  public String parameterGenerationsHBT(List<AppResult> results, List<TuningParameter> tuningParameters) {
+    Map<String, Map<String, Double>> previousUsedMetrics = extractParameterInformationIPSO(results);
+    StringBuffer idParameters = new StringBuffer();
```

Comment:
Preferably use StringBuilder instead of StringBuffer where thread safety guarantees are not required

Reply:
Will change in next PR